proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,221 @@
+# NuGetAudit without nuget.org as a package source
+<!-- Replace `Title` with an appropriate title for your design -->
+
+- Author: [zivkan](https://github.com/zivkan)
+- GitHub Issue: [12698](https://github.com/NuGet/Home/issues/12698)
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+This proposal adds a new `<auditSource>` to *NuGet.Config* files, allowing *NuGetAudit* to obtain known vulnerabilities without adding nuget.org as a package source.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+[NuGetAudit](https://learn.microsoft.com/nuget/concepts/auditing-packages) is a feature where NuGet will report during restore which packages with known vulnerabilities are being used by projects, [proposed in late 2022](../2022/vulnerabilities-in-restore.md), and finally released with .NET 8 in 2023.
+The first version of NuGetAudit used package sources as the data source for known vulnerabilities, which will be referred to as a Vulnerability Database (VDB) for the remainder of this document.
+While nuget.org provides a VDB, there are several reasons why customers may not be using nuget.org as a package source.
+This is especially true following the 2021 [blog post titled "Dependency confusion"](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) which significantly increased knowledge of multi-source package substitution supply chain risks.
+
+While the design of NuGetAudit attempts to make it easy for upstreaming package sources to also upstream the `VulnerabilityInfo` resource, the NuGet team and customers are reliant of the feed implementor to implement the change.
+An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
```

Comment:
The overlap of package source mapping and `<auditSource>` is interesting and raises some questions for me.

During the dependency confusion response, we encountered a bunch of cases where a common internal package name (internal to Microsoft or internal to 3rd party, both cases cropped up) overlapped with a NuGet.org package. With package source mapping, it allows the package consumer to say "no I don't want that NuGet.org thing, I want GenericLoggingLibrary from my own source". If the NuGet.org VDB starts reporting GenericLoggingLibrary as vulnerable, and the package consumer is using NuGet.org as an audit source, but they have a package source mapping saying GenericLoggingLibrary should come from their private feed, what should happen?

It seems like either the package source mapping would suppress the vulnerability report for that package since it's not the NuGet.org version (maybe) or it would show the warning to ere on the side of caution. It seems like the 2nd option is the safest to assume since ID conflicts between NuGet.org and remote are probably most often same content but it would be annoying for users that know the content is actually different. What would they do? Is there a good way to inform the users about this conflict and give them a suppression option, perhaps tailored to this case? It's not like they want to suppress a specific vulnerability, it's more like they want to say this package ID is not the same thing.

I guess this generalizes to the case when the user isn't using package source mapping at and only uses their internal feed, but an internal package ID just so happens to collide with a vulnerable range.

Reply:
I think you are highlighting the importance of choice here.

I should have the choice to:

- Define where my packages come from
- Define how my packages are audited
- Define the policy of how they will be audited

I don't see any of these things as concerning, rather they are headed in the right directly. We would expect that any matching package ID + version from a reputable source implementing VDB would result in someone needing to look at that dependency, right?

Reply:
On one hand, NuGet has always been based around the assumption that packages are globally unique (and immutable). While package source mapping provides a step away from the globally unique assumption, NuGet will still share the same global packages folder for a particular package id+version combination, so if solution 1 gets  package from source 1, package source mapping will not prevent source 1's package that is already in the global packages folder when solution 2 expects the package to come from source 2. So, even with package source mapping, we're not fully away from the globally unique assumption.

Additionally, the same discussion (should package source mapping scope which packages are checked by NuGet Audit) is relevant to the existing NuGet Audit functionality, where VDBs are only downloaded from package sources, even if this feature spec is not implemented.
 
This is an interesting area, however, I consider it scope creep for this feature spec. Therefore, I'm going to resolve this PR conversation, but if someone thinks it's important enough, a new issue or feature spec can be created for that specific topic.